active/0000-routeable-components.md

```diff
@@ -0,0 +1,488 @@
+resolution in a non-flat namespace.
+
+Since standalone route templates are going away in favor of
+components, the current distinction between the `templates` and
+`templates/components` directories becomes unhelpful. We want to put
+all of them into just `templates`. Thanks to the introduction of
+`<element>` and `<fragment>` (see the
+[Element and Fragment RFC](element-and-fragment)), we can make this
+transition gracefully.
+
+  - A template in `templates` with an `<element>` or `<fragment>` tag is defining a Component or Fragment.
+  - A template in `templates` with neither is a legacy standalone template.
+  - A template in `templates/components` with neither is a legacy component.
+
+### Specifying component attributes
+
+We will introduce a new route hook named `attributes`. It receives the
+route's positional parameters (the same ones that are currently passed
+to the `model` hook), plus any query parameters defined for the route,
+and must return a POJO. The POJO's properties may be promises, which
+will all be resolved in the same way we currently resolve the `model`
+hook (including invoking `loading` actions as needed). The default
+`attributes` implementation is:
+
+````js
+attributes: function(params, queryParams) {
+    return {
+        model: this.model(params)
+    }
+}
+````
+
+(Possible variation for this default implementation: any queryParams
+that are present get stuffed directly into the attributes as well,
+alongside `model`.)
+
+Unlike today's `model` hook, `attributes` gets called *every* time
+your route is entered, whether or not `transitionTo` was passed a
+preexisting model. This eliminates the need for `beforeModel` and
+`afterModel`, because you can implement either behavior from within
+your own `attributes` or `model` functions. Therefore, `beforeModel`
+and `afterModel` are deprecated.
+
+This also means we deprecate `refresh`, because routes essentially
+always refresh. 
+
+### When Routes can call render
+
+Today, `render` is normally called from the `renderTemplate` hook, but
+it can also be called from action handlers on the route. Calling
+`render` from action handlers -- outside of a route transition --
+makes routes unnecessarily stateful and introduces an additional way
+to do things that are better off done directly from within components
+and their templates. We will therefore deprecate calling `render`
+outside of transitions. See Appendix A below for an example of
+updating this old behavior.
+
+Today, `renderTemplate` receives the default controller and the
+model. But since we don't intend to have controllers anymore, we will
+deprecate `renderTemplate` in favor of `renderComponents`, which would
+have a default implementation like:
+
+````js
+renderComponents: function(attributes) {
+    this.render(this.routeName, attributes);
+}
+````
+
+The `attributes` argument comes from the new `attributes` hook as
+described in the previous section. It is resolved with `RSVP.hash`
```

Comment:
Is it possible to have an escape valve for waiting on promises?   The behavior described above forces the transition to wait on the slowest call.  Currently the way you get around this is returning a hash that contains invoked promises.  If this is the default, then maybe the escape valve is "just use an async component".